cmd/crdvalidator/test/e2e/crdvalidator_test.go

```diff
@@ -0,0 +1,197 @@
+		// CRDValidator was designed to prevent but is unknown why it is occurring. We
+		// Should come back to this test case, figure out what is preventing it from
+		// hitting the webhook and decide if we want to keep that logic or not.
+		PWhen("an incoming crd event removes a stored version", func() {
+			var crd *apiextensionsv1.CustomResourceDefinition
+
+			BeforeEach(func() {
+				crd = newTestingCRD(
+					util.GenName(defaultTestingCrdName),
+					[]apiextensionsv1.CustomResourceDefinitionVersion{
+						{
+							Name:    "v1alpha1",
+							Served:  true,
+							Storage: true,
+							Schema: &apiextensionsv1.CustomResourceValidation{
+								OpenAPIV3Schema: &apiextensionsv1.JSONSchemaProps{
+									Type:        "object",
+									Description: "my crd schema",
+								},
+							},
+						},
+						{
+							Name:    "v1alpha2",
+							Served:  true,
+							Storage: false,
+							Schema: &apiextensionsv1.CustomResourceValidation{
+								OpenAPIV3Schema: &apiextensionsv1.JSONSchemaProps{
+									Type:        "object",
+									Description: "my crd schema",
+								},
+							},
+						},
+					},
+				)
+
+				err := c.Create(ctx, crd)
+				Expect(err).To(BeNil())
+
+				Eventually(func() error {
+					return c.Get(ctx, client.ObjectKeyFromObject(crd), crd)
+				}).Should(Succeed(), "should be able to create a safe crd but was not")
+			})
+
+			AfterEach(func() {
+				By("deleting the testing crd")
+				Expect(c.Delete(ctx, crd)).To(BeNil())
+			})
+
+			It("should deny admission", func() {
+				Eventually(func() string {
+					if err := c.Get(ctx, client.ObjectKeyFromObject(crd), crd); err != nil {
+						return err.Error()
+					}
+
+					newCRD := newTestingCRD(
+						crd.Spec.Names.Singular,
+						[]apiextensionsv1.CustomResourceDefinitionVersion{
+							{
+								Name:    "v1alpha2",
+								Served:  true,
+								Storage: true,
+								Schema: &apiextensionsv1.CustomResourceValidation{
+									OpenAPIV3Schema: &apiextensionsv1.JSONSchemaProps{
+										Type:        "object",
+										Description: "my crd schema",
+									},
+								},
+							},
+							{
+								Name:    "v1alpha3",
+								Served:  true,
+								Storage: false,
+								Schema: &apiextensionsv1.CustomResourceValidation{
+									OpenAPIV3Schema: &apiextensionsv1.JSONSchemaProps{
+										Type:        "object",
+										Description: "my crd schema",
+									},
+								},
+							},
+						},
+					)
+
+					newCRD.SetResourceVersion(crd.ResourceVersion)
+					newCRD.Status.StoredVersions = []string{"v1alpha2"}
+
+					return c.Update(ctx, newCRD).Error()
+				}).Should(ContainSubstring("cannot remove stored versions"))
+			})
+		})
+	})
+})
+
+func newTestingCRD(name string, versions []apiextensionsv1.CustomResourceDefinitionVersion) *apiextensionsv1.CustomResourceDefinition {
+	return &apiextensionsv1.CustomResourceDefinition{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name + ".e2e.io",
```

Comment:
nit: I think this is slightly confusing from the consumer aspect where I pass an explicit name parameter, but it gets mutated under-the-hood. Maybe we should pass an explicit group parameter too?

Reply:
Yeah, that's a fair improvement.